src/java.base/share/classes/com/sun/crypto/provider/DH.java

```diff
@@ -0,0 +1,340 @@
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package com.sun.crypto.provider;
+
+import sun.security.util.ArrayUtil;
+import sun.security.util.CurveDB;
+import sun.security.util.ECUtil;
+import sun.security.util.Hybrid;
+import sun.security.util.NamedCurve;
+
+import javax.crypto.DecapsulateException;
+import javax.crypto.KEM;
+import javax.crypto.KEMSpi;
+import javax.crypto.KeyAgreement;
+import javax.crypto.SecretKey;
+import javax.crypto.spec.SecretKeySpec;
+import java.io.IOException;
+import java.math.BigInteger;
+import java.security.*;
+import java.security.interfaces.ECKey;
+import java.security.interfaces.ECPublicKey;
+import java.security.interfaces.XECKey;
+import java.security.interfaces.XECPublicKey;
+import java.security.spec.AlgorithmParameterSpec;
+import java.security.spec.ECPoint;
+import java.security.spec.ECPublicKeySpec;
+import java.security.spec.InvalidKeySpecException;
+import java.security.spec.KeySpec;
+import java.security.spec.NamedParameterSpec;
+import java.security.spec.XECPublicKeySpec;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+
+import static sun.security.util.SecurityConstants.PROVIDER_VER;
+
+/**
+ * The DH provider is a KEM abstraction layer over traditional DH based
+ * key exchange. It models DH/ECDH/XDH as KEMs, like post-quantum algorithms,
+ * so DH/ECDH/XDH can be used in hybrid key exchange, alongside post-quantum
+ * KEMs.
+ */
+public class DH implements KEMSpi {
+
+    // DH in its own private provider so we always getInstance from here.
+    public static final Provider PROVIDER = new ProviderImpl();
+
+    private static class ProviderImpl extends Provider {
+        @java.io.Serial
+        private static final long serialVersionUID = 0L;
+        private ProviderImpl() {
+            super("InternalJCE", PROVIDER_VER, "");
+            put("KEM.DH", DH.class.getName());
+
+            // Hybrid KeyPairGenerator/KeyFactory/KEM
+
+            // The order of shares in the concatenation for group name
+            // X25519MLKEM768 has been reversed. This is due to historical
+            // reasons.
+            var attrs = Map.of("name", "X25519MLKEM768", "left", "ML-KEM-768",
+                    "right", "X25519");
+            putService(new HybridService(this, "KeyPairGenerator",
+                    "X25519MLKEM768", "sun.security.util.Hybrid$KeyPairGeneratorImpl",
```

Comment:
Is there a reason why `Hybrid` is in `sun.security.util` instead of `com.sun.crypto.provider`?    This is the only place it's used, so this seems to be a more natural place for it, but maybe I'm just not far enough into the guts of the code.

Reply:
Agreed.